File: aavt/src/main/java/com/wuwang/aavt/media/hard/StrengthenMp4MuxStore.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.wuwang.aavt.media.hard;

import android.media.MediaFormat;
import android.media.MediaMuxer;

import com.wyz.common.api.IHardStore;
import com.wyz.common.core.base.HardMediaData;

import java.io.IOException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * StrengthenMp4MuxStore
 *
 * @author wuwang
 * @version v1.0 2017:11:08 17:15
 */
public class StrengthenMp4MuxStore implements IHardStore {

    private MediaMuxer mMuxer;
    private final boolean av;
    private String path;
    private int audioTrack = -1;
    private int videoTrack = -1;
    private final Object Lock = new Object();
    private boolean muxStarted = false;
    private LinkedBlockingQueue<HardMediaData> cache;
    private Recycler<HardMediaData> recycler;
    private ExecutorService exec;

    public StrengthenMp4MuxStore(boolean av) {
        this.av = av;
        cache = new LinkedBlockingQueue<>(30);
        recycler = new Recycler<>();
        exec = new ThreadPoolExecutor(1, 1, 1, TimeUnit.MINUTES,
                new LinkedBlockingQueue<Runnable>(16), Executors.defaultThreadFactory());
    }

    @Override
    public void close() throws Exception {
        synchronized (Lock) {
            if (muxStarted) {
                audioTrack = -1;
                videoTrack = -1;
                muxStarted = false;
            }
        }
    }

    private void muxRun() {
        while (muxStarted) {
            try {
                HardMediaData data = cache.poll(1, TimeUnit.SECONDS);
                synchronized (Lock) {
                    if (muxStarted && data != null) {
                        mMuxer.writeSampleData(data.getIndex(), data.getData(), data.getInfo());
                        recycler.put(data.getIndex(), data);
                    }
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            mMuxer.stop();
            mMuxer.release();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
        mMuxer = null;
        cache.clear();
        recycler.clear();
    }

    @Override
    public int addTrack(MediaFormat mediaFormat) {
        int ret = -1;
        synchronized (Lock) {
            if (!muxStarted) {
                if (audioTrack == -1 && videoTrack == -1) {
                    try {
                        mMuxer = new MediaMuxer(path, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                String mime = mediaFormat.getString(MediaFormat.KEY_MIME);
                if (mime.startsWith("audio")) {
                    audioTrack = mMuxer.addTrack(mediaFormat);
                    ret = audioTrack;
                } else if (mime.startsWith("video")) {
                    videoTrack = mMuxer.addTrack(mediaFormat);
                    ret = videoTrack;
                }
                startMux();
            }
        }
        return ret;
    }

    private void startMux() {
        boolean canMux = !av || (audioTrack != -1 && videoTrack != -1);
        if (canMux) {
            mMuxer.start();
            muxStarted = true;
            exec.execute(new Runnable() {
                @Override
                public void run() {
                    muxRun();
                }
            });
        }
    }

    @Override
    public int addData(int track, HardMediaData hardMediaData) {
        if (track >= 0) {
            hardMediaData.setIndex(track);
            if (track == audioTrack || track == videoTrack) {
                HardMediaData d = recycler.poll(track);
                if (d == null) {
                    d = hardMediaData.copy();
                } else {
                    hardMediaData.copyTo(d);
                }
                while (!cache.offer(d)) {
                    HardMediaData c = cache.poll();
                    recycler.put(c.getIndex(), c);
                }
            }
        }
        return 0;
    }

    @Override
    public void setOutputPath(String path) {
        this.path = path;
    }

}

File: common/src/main/java/com/wyz/common/core/mux/StrengthenMp4MuxStore.kt
package com.wyz.common.core.mux

import android.media.MediaFormat
import android.media.MediaMuxer
import com.wyz.common.api.IHardStore
import com.wyz.common.core.base.HardMediaData
import java.io.IOException
import java.util.concurrent.*

class StrengthenMp4MuxStore : IHardStore {
    private var mMuxer: MediaMuxer? = null
    private val av : Boolean
    private var path: String? = null

    private var audioTrack = -1
    private var videoTrack = -1

    private val Lock = Object()

    private var muxStarted = false

    private val cache: LinkedBlockingQueue<HardMediaData>
    private var recycler: Recycler<HardMediaData>? = null
    private var exec: ExecutorService? = null

    constructor(av: Boolean) {
        this.av = av
        cache = LinkedBlockingQueue(30)
        recycler = Recycler()
        exec = ThreadPoolExecutor(1, 1, 1, TimeUnit.MINUTES, LinkedBlockingQueue(16), Executors.defaultThreadFactory())
    }

    override fun setOutputPath(path: String) {
        this.path = path
    }

    override fun addTrack(track: MediaFormat): Int {
        var ret = -1
        val s = path ?: return ret
        synchronized(Lock) {
            if (!muxStarted) {
                if (audioTrack == -1 && videoTrack == -1) {
                    try {
                        mMuxer = MediaMuxer(s, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
                mMuxer?.also {
                    val mime= track.getString(MediaFormat.KEY_MIME) ?:return ret
                    if (mime.startsWith("audio")) {
                        audioTrack = it.addTrack(track)
                        ret = audioTrack
                    } else if (mime.startsWith("video")) {
                        videoTrack = it.addTrack(track)
                        ret = videoTrack
                    }
                    startMux()
                }
            }
        }
        return ret
    }

    override fun addData(track: Int, data: HardMediaData): Int {
        if (track >= 0) {
            data.index = track
            if (track == audioTrack || track == videoTrack) {
                var d = recycler?.poll(track)
                if (d == null) {
                    d = data.copy()
                } else {
                    data.copyTo(d)
                }
                while (!cache.offer(d)) {
                    val c = cache.poll() ?:continue
                    recycler?.put(c.index, c)
                }
            }
        }
        return 0
    }

    private fun startMux() {
        val canMux = !av || (audioTrack != -1 && videoTrack != -1)
        if (canMux) {
            mMuxer?.start()
            muxStarted = true
            exec?.execute { muxRun() }
        }
    }

    private fun muxRun() {
        while (muxStarted) {
            try {
                val data = cache.poll(1, TimeUnit.SECONDS)
                synchronized(Lock) {
                    if (muxStarted && data != null) {
                        mMuxer?.writeSampleData(data.index, data.data, data.info)
                        recycler?.put(data.index, data)
                    }
                }
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        try {
            mMuxer?.stop()
            mMuxer?.release()
            mMuxer = null
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
        cache.clear()
        recycler?.clear()
    }

    override fun close() {
        synchronized(Lock) {
            if (muxStarted) {
                audioTrack = -1
                videoTrack = -1
                muxStarted = false
            }
        }
    }
}